libdnf/utils/exception.cpp

```diff
@@ -0,0 +1,61 @@
+/*
+Copyright (C) 2021 Red Hat, Inc.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "libdnf/utils/exception.hpp"
+
+#include <charconv>
+#include <system_error>
+
+
+namespace libdnf {
+
+
+const char * Exception::what() const noexcept {
+    try {
+        str_what = std::string(get_domain_name()) + "::" + get_name() + ": " + get_description();
+
+        // related data, eg file name
+        const char * what = runtime_error::what();
+        if (what && what[0] != '\0') {
+            str_what += ": ";
+            str_what += what;
+        }
+
+        return str_what.c_str();
+    } catch (...) {
+        return runtime_error::what();
+    }
+}
+
+const char * SystemError::get_name() const noexcept {
+    std::to_chars(name + 6, name + sizeof(description) - 1, get_error_code());
```

Comment:
It would appear you meant to use `sizeof(name)` here?

But the number shouldn't be part of the name anyway. It should be part of the description and assumption is we are composing the description from multiple strings anyway, so I'd drop this low-level approach and just join the strings the usual way.

Reply:
> It would appear you meant to use sizeof(name) here?

Of course. There must be `sizeof(name)` there. I will fix it. Thanks.

> But the number shouldn't be part of the name anyway.

Hmm. I don't want all system errors to have the same name.

Reply:
> Hmm. I don't want all system errors to have the same name.

Why? It adds unnecessary complexity, I don't see the benefit. What matters is the error messages are complete information-wise. If someone will want to handle different `errno`s differently, they will base the condition on that (FTR I think the `get_error_code()` method should be called `get_errno()`, because that's the name everyone is used to, if it's called `get_error_code()`, it's unnecessarily misleading). I'm not actually sure what the name could be used for, but I think it's just fine without the number.